### PR TITLE
ci: shard the `clang-tidy` build

### PR DESCRIPTION
**Pull request by coryan**
Reduce the worst-case build time by (imperfectly) sharding the build N ways.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12966)

### CODE REVIEW

**Review by cum1997 — APPROVED** [review]
Troubleshooting info:
  Principal: nguyendangduc811997@gmail.com
  Resource: 936212892354
  Troubleshooting URL: console.cloud.google.com/iam-admin/troubleshooter;permissions=cloudbuild.builds.get;principal=nguyendangduc811997@gmail.com;resources=%2F%2Fcloudresourcemanager.googleapis.com%2Fprojects%2F936212892354/result

Missing permissions:
  cloudbuild.builds.get

File: ci/lib/shard.sh [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+)
+
+readonly BIGQUERY_SHARD=(
+  bigquery
+  experimental-bigquery_rest
+)
+
+readonly PUBSUB_SHARD=(
+  pubsub
+  # Pub/Sub Lite is included because has hand-crafted code and requires
```
Comment (dbolduc) [inline ci/lib/shard.sh:48]: s/because has/because it has/

File: ci/cloudbuild/triggers/clang-tidy-compute.yaml [hunk 2]
```diff
@@ -13,9 +13,10 @@ sourceToBuild:
   repoType: GITHUB
   uri: https://github.com/googleapis/google-cloud-cpp
 substitutions:
-  _BUILD_NAME: clang-tidy-compute
+  _BUILD_NAME: clang-tidy
```
Comment (dbolduc) [inline ci/cloudbuild/triggers/clang-tidy-compute.yaml:16]: comment: It should be fine to reuse this output directory.

https://github.com/googleapis/google-cloud-cpp/blob/d807b786a8300f128d782fb7e5e2342c41725247/ci/cloudbuild/build.sh#L297

File: ci/cloudbuild/cloudbuild.yaml [hunk 3]
```diff
@@ -104,6 +105,7 @@ steps:
   env: [
     'BAZEL_REMOTE_CACHE=https://storage.googleapis.com/${_CACHE_BUCKET}/bazel-cache/${_DISTRO}-${_BUILD_NAME}',
     'LIBRARIES=${_LIBRARIES}',
```
Comment (dbolduc) [inline ci/cloudbuild/cloudbuild.yaml:107]: comment: `LIBRARIES` and `SHARD` seem to be serving the same purpose. One is for publish-docs, the other is for clang-tidy.

Having two variables seems non-ideal... but I am not the one doing the work to make the repo a nicer place to live :shrug:

**Review by dbolduc — APPROVED** [review]

File: ci/lib/shard.sh (hunk 1 above)
Comment (coryan) [reply ci/lib/shard.sh]: Fixed.

File: ci/cloudbuild/triggers/clang-tidy-compute.yaml (hunk 2 above)
Comment (coryan) [reply ci/cloudbuild/triggers/clang-tidy-compute.yaml:16]: Thanks for catching that, fixed now.

File: ci/cloudbuild/cloudbuild.yaml (hunk 3 above)
Comment (coryan) [reply ci/cloudbuild/cloudbuild.yaml:107]: Yes.  The sharding for the documents is very ad-hoc.  I am going to leave it alone.

File: ci/cloudbuild/triggers/clang-tidy-compute.yaml (hunk 2 above)
Comment (dbolduc) [reply ci/cloudbuild/triggers/clang-tidy-compute.yaml:16]: So it wasn't fine?

Comment (coryan) [reply ci/cloudbuild/triggers/clang-tidy-compute.yaml:16]: It does not work right if you try to compile two shards locally.

**Pull request merged**